core/embed/extmod/modtrezorcrypto/modtrezorcrypto-bip32.h

```diff
@@ -266,15 +266,18 @@ STATIC mp_obj_t mod_trezorcrypto_HDNode_serialize_public(mp_obj_t self,
                                                          mp_obj_t version) {
   uint32_t ver = trezor_obj_get_uint(version);
   mp_obj_HDNode_t *o = MP_OBJ_TO_PTR(self);
-  char xpub[XPUB_MAXLEN] = {0};
+  vstr_t xpub = {0};
+  vstr_init_len(&xpub, XPUB_MAXLEN);
```

Comment:
Here (and on a few other places) you allocate a buffer that is one byte longer than you actually need, since `xpub.alloc = XPUB_MAXLEN + 1`. That's completely OK; however, I'm not sure whether you did it intentionally.

Reply:
I think that's not quite true. Here we allocate 16 more bytes than we actually need and similarly in some other places. I can do some research and change the values of the constants to make them tight, but I think they were intentionally chosen as round numbers.

However, on a different note, it may be worth rounding the constants to one less than a power of two or one less than a multiple of four. So for example we can change `XPUB_MAXLEN` to 112, 115 or 127. Opinions?

Reply:
I am OK with merging this PR as it is.

As for constants, I think:
a) we should move them to trezor-crypto
b) not round them and use the them tight, but as Andrew pointed out this requires lots of effort and research, so I am fine with keeping them as they are

Reply:
I already did the research into `XPUB_MAXLEN` (spec + quick look at our code), so we might as well use it in accordance with (b). There are a few places where I allocated one more byte than we used to and where it's possibly unnecessary, which is what Ondra was referring to, but I didn't change the numbers because it seemed they were intentionally rounded. What I hadn't thought of, however, is that the allocator most likely aligns to multiples of 4. So it makes more sense to prefer rounding them to one less than a multiple of 4. The only other place that I am actively aware of is `ADDRESS_MAXLEN`, so I am changing that to 39. There might be other places, like in Monero, but I am not 100% sure about those and doesn't seem worth investigating.

Reply:
As for (a), I moved the 2 constants to bip32.h and used them wherever it was possible: https://github.com/trezor/trezor-firmware/pull/1305/commits/49eaccd94bfb0ee807157cc423e33d05ea00565a